### ISSUE TITLE
Issues Using Anything but Default ASBD for EZMicrophone

### ISSUE THREAD

**Issue by vfxdrummer** [syedhali/EZAudio#239]
Hey Syed -

First of all, amazing framework! I love it. I was looking for a way to capture the iOS microphone audio data and EZMicrophone definitely does the job. One thing, though. It seems that when I change AudioStreamBasicDescription to anything but the default, I receive several errors like :

Error: Failed to fill complex buffer in float converter ('insz') 

Here are example settings :
audioStreamBasicDescription.mFormatID = kAudioFormatLinearPCM;
audioStreamBasicDescription.mSampleRate = 16000;
audioStreamBasicDescription.mFramesPerPacket = 1;
audioStreamBasicDescription.mBytesPerPacket = 2; //16 bits * 1 channel
audioStreamBasicDescription.mBytesPerFrame = 2;
audioStreamBasicDescription.mChannelsPerFrame = 1; //1 channel
audioStreamBasicDescription.mBitsPerChannel = 16;
audioStreamBasicDescription.mFormatFlags = kAudioFormatFlagIsSignedInteger |
kAudioFormatFlagIsBigEndian | kAudioFormatFlagIsPacked;
audioStreamBasicDescription.mReserved = 0;

Basically, I am trying to make the stream 16 bits / 16 kHz so that I can upload it to Nuance's speech to text dictation.

Any thoughts on this? Thanks again!

**Comment by vfxdrummer** [comment]
I also tried creating my ASBD using 
 audioStreamBasicDescription = [EZAudioUtilities monoFloatFormatWithSampleRate:16000];

But, that produces errors, as well.

I am only using EZMicrophone. I need 16 bit / 16 kHz pcm audio data in order to POST it to the Nuance speech-To-Text dictation API. We were saving a wav file and uploading it which works, but has a lag.

Should I just convert the 32-bit / 44.1kHz pcm data that I get from the AudoBufferList in the microphone delegate? Any thoughts or pointers?

Thanks again for this great framework!

-Tim

**Comment by vfxdrummer** [comment]
Hey - 

Quick Update after scouring the interwebs. I got it to work. This was the missing piece :

  audioStreamBasicDescription.mFormatFlags = kAudioFormatFlagIsSignedInteger | kAudioFormatFlagsNativeEndian | kAudioFormatFlagIsPacked;

After looking at this code snippet from 'The Amazing Audio Engine' :

//  The Amazing Audio Engine
//  .mFormatID = kAudioFormatLinearPCM,
//  .mFormatFlags = (sampleType == AEAudioStreamBasicDescriptionSampleTypeFloat32
//                   ? kAudioFormatFlagIsFloat : kAudioFormatFlagIsSignedInteger | kAudioFormatFlagsNativeEndian)
//  | kAudioFormatFlagIsPacked
//  | (interleaved ? 0 : kAudioFormatFlagIsNonInterleaved),
//  .mChannelsPerFrame  = numberOfChannels,
//  .mBytesPerPacket    = sampleSize * (interleaved ? numberOfChannels : 1),
//  .mFramesPerPacket   = 1,
//  .mBytesPerFrame     = sampleSize * (interleaved ? numberOfChannels : 1),
//  .mBitsPerChannel    = 8 * sampleSize,
//  .mSampleRate        = sampleRate,

I realized that 16 bit data is SInt16 and needs these format flags :

  audioStreamBasicDescription.mFormatFlags = kAudioFormatFlagIsSignedInteger | kAudioFormatFlagsNativeEndian | kAudioFormatFlagIsPacked;

And, I retrieve it from the buffer like so :
  AudioBuffer audioBuffer = bufferList->mBuffers[0];

  SInt16 *frame = (SInt16*)audioBuffer.mData;

  [self.audioData appendBytes:frame length:audioBuffer.mDataByteSize];

Works for me - thanks again for the great framework!

**Comment by Weezlo** [comment]
@vfxdrummer Thanks for your update Tim ! you saved me from some really bad headaches ! (i had quite a lot aleady)
I was casting mData to Floats instead of Signed Ints, that was my mistake...
By the way, i'm also sending audio data to Nuance Speech Recognition API :)
I'm wondering, are you streaming audio speech to a remote server and doing real time transcription with it ? For the moment I send the whole Wav file once recorder but i was also thinking about streaming data. What API are you using? SpeechMagic SDK?

Thank you again.

Anis
